### PR TITLE
TaskActivityDispatcher -> set activity to current before middleware

### PR DESCRIPTION
**Pull request by jviau**
Quick change to set the current activity _before_ running the middleware pipeline. This allows for accessing `Activity.Current` during the middleware. Otherwise it is `null` until the `TaskActivity` is actually executed.

### CODE REVIEW

File: src/DurableTask.Core/TaskActivityDispatcher.cs [hunk 1]
```diff
@@ -148,11 +149,15 @@ async Task OnProcessWorkItemAsync(TaskActivityWorkItem workItem)
...
                 dispatchContext.SetProperty(scheduledEvent);
 
-                await this.dispatchPipeline.RunAsync(dispatchContext, async _ =>
+                // correlation
+                CorrelationTraceClient.Propagate(() =>
                 {
-                    // correlation 
-                    CorrelationTraceClient.Propagate(() => workItem.TraceContextBase?.SetActivityToCurrent());
+                    workItem.TraceContextBase?.SetActivityToCurrent();
+                    diagnosticActivity = workItem.TraceContextBase?.CurrentActivity;
```
Comment (cgillum) [inline src/DurableTask.Core/TaskActivityDispatcher.cs:156]: Looks like you're not using this local variable for anything?

Comment (jviau) [reply src/DurableTask.Core/TaskActivityDispatcher.cs:156]: Line 207 I am stopping the activity - to make sure it is removed from `Activity.Current` before exiting this function.

**Comment by cgillum** [comment]
@bachuv or @TsuyoshiUshio, would either of you be able to sign off on this change?

File: src/DurableTask.Core/TaskActivityDispatcher.cs [hunk 2]
```diff
@@ -148,11 +149,15 @@ async Task OnProcessWorkItemAsync(TaskActivityWorkItem workItem)
...
+                });
 
+                await this.dispatchPipeline.RunAsync(dispatchContext, async _ =>
+                {
```
Comment (bachuv) [inline src/DurableTask.Core/TaskActivityDispatcher.cs:160]: @jviau Thank you for contributing! Can you share some scenarios or sample code to explain these changes and how they would be helpful?

Comment (jviau) [reply src/DurableTask.Core/TaskActivityDispatcher.cs:160]: As mentioned in the PR description, this sets `Activity.Current` _before_ the middleware pipeline is executed, so it is accessible during user provided middleware (`TaskHubWorker.AddActivityDispatcherMiddleware`). Having Activity.Current set there is important so any telemetry during your middleware is also correlated to TaskActivity that is about to be executed. Without this change, nothing in the middleware during a TaskActivity can be correlated. However, TaskOrchestration's middleware does not face the same issue - correlation is available there.

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum and @bachuv 

The change looks good. I'd like to run the following test and see if it works.  How can we run the test?
 https://github.com/Azure/durabletask/blob/main/test/DurableTask.AzureStorage.Tests/Correlation/CorrelationScenarioTest.cs

**Comment by jviau** [comment]
@cgillum - do you know why the build failed there? It shows the tests as 'Flaky'?

**Review by bachuv — APPROVED** [review]

**Pull request merged**